src/DynamoCore/Utilities/LuceneSearchUtility.cs

```diff
@@ -36,23 +46,34 @@ internal LuceneSearchUtility(DynamoModel model)
         /// <summary>
         /// Initialize Lucene config file writer.
         /// </summary>
-        internal void InitializeLuceneConfig(string dirName)
+        internal void InitializeLuceneConfig(string dirName, LuceneStorage storageType = LuceneStorage.FILE_SYSTEM)
         {
             addedFields = new List<string>();
 
-            DirectoryInfo webBrowserUserDataFolder;
+            DirectoryInfo luceneUserDataFolder;
             var userDataDir = new DirectoryInfo(dynamoModel.PathManager.UserDataDirectory);
-            webBrowserUserDataFolder = userDataDir.Exists ? userDataDir : null;
+            luceneUserDataFolder = userDataDir.Exists ? userDataDir : null;
 
             directory = dirName;
-            string indexPath = Path.Combine(webBrowserUserDataFolder.FullName, LuceneConfig.Index, dirName);
-            indexDir = Lucene.Net.Store.FSDirectory.Open(indexPath);
+            string indexPath = Path.Combine(luceneUserDataFolder.FullName, LuceneConfig.Index, dirName);
+
+            currentStorageType = storageType;
+
+            if (storageType == LuceneStorage.RAM)
+            {
+                indexDir = new RAMDirectory();
+            }
+            else
+            {
+                indexDir = FSDirectory.Open(indexPath);
+            }
+                
 
             // Create an analyzer to process the text
             Analyzer = new StandardAnalyzer(LuceneConfig.LuceneNetVersion);
 
-            // Initialize Lucene index writer, unless in test mode.
-            if (!DynamoModel.IsTestMode)
+            // Initialize Lucene index writer, unless in test mode or we are using RAMDirectory for indexing info.
+            if (!DynamoModel.IsTestMode || currentStorageType == LuceneStorage.RAM)
```

Comment:
Why do we need this condition `currentStorageType == LuceneStorage.RAM` here, when we are not using any files to index

Reply:
yes, we need the condition otherwise a specific test is failing.
Basically I've added this validation because in the test SearchNodeAutocompletionSuggestions() is doing th next actions:

- Populate Node Autocomplete with candidates
- Search the text "ar" using Node Autocomplete Search (over a small set of node and using Lucene Search with RAMDirectory).
- Check that we are getting 5 results.

So if you check this piece of code is creating the writer then if IsTestMode is True (due that is executing the text) then the writer won't be created and when searching it will return null or crash (meaning that no results were found), the same case for the validation added in the method InitializeIndexDocumentForNodes(), if we don't check the currentStorageType variable then is always returning null and is crashing.

Just to clarify things: remember that with NodeAutocomplete indexing process we are using RAMDirectory (just for the subset of nodes) so the IsTestMode flag doesn't matter any more (there is no problem with concurrency and will be executing the search over the subset of nodes indexed), that's why we have to check for the currentStorageType variable.

At the end we will be removing all those conditions related to IsTestMode and currentStorageType but we need to refactor the code so that when tests are being executed then the indexing is done using RAMDirectory otherwise we will be using FSDirectory.

Please let me know if this explanation is clear or we need to check the implementation in detail

Thanks

Reply:
Thanks I think that matches my expectation, we should refactor the code soon for unit testing